fix: Permission check failing when privilege is at ROLE level

sergio-ropero opened the pull request:
# What
The replication can be only done when the user has `SUPERUSER` or `REPLICATION` privileges. 
```
psql: error: FATAL:  must be superuser or replication role to start walsender
```
In the documentation we assumed that using a `role` with any of the above privileges attached to a user, grants these attributes to the user, [but it's not right](https://www.postgresql.org/docs/current/role-attributes.html#:~:text=initiating%20replication,REPLICATION%20LOGIN).

# Solution
In order to fix the check function we will try directly to connect to the database doing a replication connection and verify that it's correct.

Added tests for all cases:
- Access with SUPERUSER privilege
- Access with REPLICATION privilege
- NO access (no privileges)


# FROM HERE IS NOT RELEVANT ANYMORE! (But can be interesting to keep)
This commit fixes the issue when permission is granted at ROLE level instead of USER level.

Missing revoke privileges in the tests.

## These are the test and user configuration
### Set up
The user doesn't have replication privilege at user level, neither for the role attached (`secondary_role`). But the role `secondary_role` is attached to `airbyte_replication` role, so at the end the user has access to replication thanks to last link.
```
CREATE USER airbyte_ro PASSWORD 'XXXXX';
CREATE ROLE airbyte_replication REPLICATION LOGIN;
CREATE ROLE secondary_role LOGIN;
GRANT secondary_role TO airbyte_ro;
GRANT secondary_role TO airbyte_replication;
```
### User Table
We see it doesn't have replication permission:
```
select usesysid, usename, userepl from pg_user where usename = 'airbyte_ro';
```
<img width="377" alt="Screenshot 2023-02-16 at 16 54 56" src="https://user-images.githubusercontent.com/42538006/219418768-47719ea5-ae80-495d-a8dd-ed8af30b7a83.png">

## Assigned roles
As well without replication permission.
```
select pgr.rolname, rolreplication  from pg_catalog.pg_roles pgr
join pg_catalog.pg_auth_members pgam on pgam.roleid = pgr.oid 
join pg_catalog.pg_user pgu on pgam.member = pgu.usesysid 
where usename ='airbyte_ro'; 
```
<img width="263" alt="Screenshot 2023-02-16 at 16 56 19" src="https://user-images.githubusercontent.com/42538006/219419214-92860b55-fd46-4698-acb3-7b3ac9b633e0.png">

### Query to check permission (by any subrole).
If it return at least one record, the user has privilege to Postgres replication:
```
WITH RECURSIVE cte AS (
   SELECT usesysid  FROM pg_catalog.pg_user WHERE usename = 'airbyte_ro'
   UNION ALL
   SELECT m.roleid
   FROM   cte
   JOIN   pg_auth_members m ON m.member = cte.usesysid
   )
SELECT rolname FROM cte
JOIN pg_roles ON usesysid = oid
WHERE rolreplication;
```
<img width="133" alt="Screenshot 2023-02-16 at 17 01 36" src="https://user-images.githubusercontent.com/42538006/219420611-59ca1c1c-4688-4e3e-ba4c-abe12b47629f.png">

Commit:
This commit fixes the issue when permission is granted at ROLE level instead of USER level.

Missing revoke privileges in the tests.

Commit:
Change the query to recursively look for all roles asigned.

Also improve testing.

sergio-ropero commented:
/test connector=connectors/source-postgres

rodireich commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:
```diff
@@ -322,14 +322,34 @@ public List<CheckedConsumer<JdbcDatabase, Exception>> getCheckOperations(final J
+        // Verify if the user has privilege at role level
+        final List<JsonNode> rolePrivileges = database.queryJsons(connection -> {
+          final PreparedStatement ps = connection.prepareStatement("""
+              WITH RECURSIVE cte AS (
```
So looking recursively what roles the assigned roles have is what is needed?

sergio-ropero replied:
Yes, basically if we have:
granted roleA (No replication) to user
granted roleB (No replication) to roleA
granted roleC (Replication) to roleB

The query will output roleC, so we are sure that a subrole has access.
I added the test to cover that case (only one lvl BTW)

sergio-ropero replied:
Adding extra comment in the function

Commit:
Add test for subrole with replication access

Commit:
formatting

sergio-ropero commented:
@rodireich @subodh1810 @bleonard 
The failures in the `/test` command are related to the acceptance test that we know about:
```
FAILED test_incremental.py::TestIncremental::test_two_sequential_reads[inputs0]
FAILED test_incremental.py::TestIncremental::test_two_sequential_reads[inputs1]
```
Can I just move forward and publish skipping the tests?
Also, once merged, I will change the mask in Airbyte cloud to get >= latest version (1.0.46 if I'm not wrong).
Thanks all!

sergio-ropero commented:
/test connector=connectors/source-postgres

Commit:
Roles don't share attributes, only accesses.
That means that the REPLICATION or SUPERUSER can not be inhered to the user. Because of that, we need to make the user have REPLICATION access directly.

Commit:
Roles don't share attributes, only accesses.
That means that the REPLICATION or SUPERUSER can not be inhered to the user. Because of that, we need to make the user have REPLICATION access directly.

Commit:
Bump versions and update alloydb docs

rodireich commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:
```diff
@@ -318,18 +318,18 @@ public List<CheckedConsumer<JdbcDatabase, Exception>> getCheckOperations(final J
         }, sourceOperations::rowToJson);
 
-        if (!userPrivileges.get(0).get("userepl").asBoolean()) {
+        if (!userPrivileges.get(0).get("has_access").asBoolean()) {
```
Do we know whether the cloud users that had the problem with 1.0.45 will have this permission?

sergio-ropero replied:
I have asked to the user to understand how they set up. Based on that I will adapt the PR, but this should be enough (supposedly)

rodireich commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:
```diff
@@ -318,18 +318,18 @@ public List<CheckedConsumer<JdbcDatabase, Exception>> getCheckOperations(final J
       });
 
-      // Verify that the db user has required privilege to perform replication.
+      // Verify that the db user has required privilege to perform replication. The user has to be super user of to has replication access.
```
super user of db?

sergio-ropero replied:
My bad .. I will modify the comment.
Basically the db user needs to be superuser or to have replication privilege.

Commit:
improve comment

Commit:
Typo

sergio-ropero commented:
/test connector=connectors/source-postgres

sergio-ropero commented:
/test connector=connectors/source-postgres-strict-encrypt

sergio-ropero commented:
/test connector=connectors/source-alloydb

sergio-ropero commented:
/test connector=connectors/source-alloydb-strict-encrypt

subodh1810 reviewed (commented):
The more I think about the permission checks, the more I feel that we should just instead establish a replication connection to the source DB and run the `IDENTIFY_SYSTEMS` query to make sure that the connection is valid. If the user has right permissions it will work otherwise it would fail.

Commit:
remove airbyte-oauth (#23258)

* remove airbyte-oauth

* remove oauth usages

* delete the oauth folder

Commit:
Remove airbyte persistence, analytics, and metrics (#23263)

* remove airbyte-metrics

* format

* fix pmd failures

* remove airbyte-persistence

* remove airbyte-analytics

* format

Commit:
Change from checking the permisions in `pg_users` to execute `createConnection` and verifying the connection is fine for CDC.

Commit:
Remove unneeded import

perangel committed:
Ensure PR branch is up to date with master (#23274)

* ci: Ensure PR branch is up to date with master

* Use ubuntu-latest to fix deprecation warning

perangel committed:
Use origin/master instead of master

Commit:
format

perangel committed:
ci: set user.email and user.name

perangel committed:
ci: remove command dispatch guard

perangel committed:
ci: fix merge strategy to use `-s org -X ours`

sergio-ropero commented:
/test connector=connectors/source-postgres

Commit:
Merge remote-tracking branch 'origin/master' into fix/potgres-source-permission-check

evantahler commented:
There's a lot going on here... perhaps a rebase/recreate from master might be needed after the monorepo changes

sergio-ropero commented:
> 

Yeah... I mess up a bit because this branch was created from master and being updated from there frequently. Then I rebased if from this integration branch (`dbdw-srcs-integration-code-freeze-feb23`) which allows us to run the tests for postgres.
In that rebase process I had some headaches because of master commits that were not in this branch, but I sorted it. I think it's fine now:
<img width="695" alt="Screenshot 2023-02-21 at 17 56 45" src="https://user-images.githubusercontent.com/42538006/220410188-b8504d73-6199-41d0-98a4-3a27fdbde678.png">

subodh1810 commented on airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/PostgresDebeziumStateUtil.java:
```diff
@@ -111,8 +90,8 @@ public void commitLSNToPostgresDatabase(final JsonNode jdbcConfig,
-    try (final BaseConnection pgConnection = ((BaseConnection) connection(jdbcConfig))) {
-      validateReplicationConnection(pgConnection);
+    try {
+      final BaseConnection pgConnection = (BaseConnection) PostgresDebeziumConnection.createConnection(jdbcConfig);
```
this should be in a try block otherwise the connection will not be closed

subodh1810 commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSource.java:
```diff
@@ -318,21 +320,14 @@ public List<CheckedConsumer<JdbcDatabase, Exception>> getCheckOperations(final J
-
-        if (!userPrivileges.get(0).get("userepl").asBoolean()) {
+        try {
+          PostgresDebeziumConnection.createConnection(config);
```
We need to close the connection as well so we can put this in try with resources block

subodh1810 reviewed (changes requested):
We are not closing the replication connection which can cause issues.

sergio-ropero commented:
Closing and creating a brand new PR
